clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,246 @@
+  | First_Word_Fall_Through
+  deriving (Show, Generic)
+
+data DcImplementation
+  = Independent_Clocks_Block_RAM
+  | Independent_Clocks_Distributed_RAM
+  | Independent_Clocks_Builtin_FIFO
+  deriving (Show, Generic)
+
+data DcConfig depth = DcConfig
+  { dcDepth          :: !depth
+  , dcImplementation :: !DcImplementation
+  , dcReadMode       :: !ReadMode
+  , dcReadDataCount  :: !Bool
+  , dcWriteDataCount :: !Bool
+  , dcOverflow       :: !Bool
+  , dcUnderflow      :: !Bool
+  }
+  deriving (Show, Generic)
+
+defConfig :: KnownNat depth => DcConfig (SNat depth)
+defConfig = DcConfig
+  { dcDepth = SNat
+  , dcImplementation = Independent_Clocks_Block_RAM
+  , dcReadMode = Standard_FIFO
+  , dcReadDataCount = True
+  , dcWriteDataCount = True
+  , dcOverflow = False
+  , dcUnderflow = False
+  }
+
+data FifoState n = FifoState
+  { hsQueue      :: Seq.Seq (BitVector n)
+  , relativeTime :: Int
+  } deriving Show
+
+data XilinxFifo read write depth n =
+  XilinxFifo
+    { writeReset :: Signal write ResetBusy
+    , isFull :: Signal write Full
+    , isOverflow :: Signal write Bool
+    , writeCount :: Signal write (DataCount depth)
+    , readReset :: Signal read ResetBusy
+    , isEmpty :: Signal read Empty
+    , isUnderflow :: Signal read Bool
+    , readCount :: Signal read (DataCount depth)
+    , fifoData :: Signal read (BitVector n)
+    }
+
+-- | Mock Xilinx FIFO, see [documentation](https://docs.xilinx.com/v/u/en-US/pg057-fifo-generator)
+-- for expected behavior.
+--
+-- If 'dcReadDataCount' or 'dcWriteDataCount' are disabled, the relevant signals
+-- will be 'deepErrorX'
+dcFifo ::
+  forall depth n write read .
+  ( KnownNat n
+  , KnownDomain write
+  , KnownDomain read
+
+  , KnownNat depth
+  -- Number of elements should be between [2**4, 2**17] ~ [16, 131072].
+  , 4 <= depth
+  , depth <= 17
+  ) =>
+  DcConfig (SNat depth) ->
+
+  Clock write -> Clock read ->
+  -- | Asynchronous reset
+  Reset read ->
+
+  -- | Write data
+  Signal write (Maybe (BitVector n)) ->
+  -- | Read enable
+  Signal read Bool ->
+  XilinxFifo read write depth n
+dcFifo DcConfig{..} wClk rClk rst writeData rEnable =
+  let
+    (wRstBusy, wFull, wOver, wCnt, rRstBusy, rEmpty, rUnder, rCnt, rData) =
+      go initState rstSignalR rEnable rstSignalW writeData
+  in XilinxFifo
+      wRstBusy
+      wFull
+      (if dcOverflow then wOver else deepErrorX "Overflow disabled")
+      (if dcWriteDataCount then wCnt else deepErrorX "Write data count disabled")
+      rRstBusy
+      rEmpty
+      (if dcUnderflow then rUnder else deepErrorX "Underflow disabled")
+      (if dcReadDataCount then rCnt else deepErrorX "Read data count disabled")
+      (deepErrorX "No sample" :- rData)
+
+ where
+  rstSignalR = unsafeToHighPolarity rst
+  rstSignalW = unsafeSynchronizer rClk wClk $ unsafeToHighPolarity rst
+
+  -- reified depth
+  maxDepth = snatToNum @Int (powSNat (SNat @2) (SNat @depth))
+
+  -- TODO: how do we ensure it doesn't "space leak" in Haskell?
```

Comment:
That's a good question. Typically what we do is strategically add `seqX` statements. You should ensure that evaluating an element from one signal evaluates elements from all other signals too. This prevents thunk buildups when a user only uses one signal.

The blockram / register implementations should give you some inspiration.

Reply:
Ok, I think based on inspection `seqX` will not be needed? I also ran simulation a little bit and that didn't blow up with `-K1K` so I'm pretty confident there is no thunk leak.